### PR TITLE
Add autoscroll feature to flag views

### PR DESCRIPTION
**Pull request by pelumy**
Final part of [#1489](https://github.com/planetary-social/nos/issues/1489) and [#1493](https://github.com/planetary-social/nos/issues/1493)

## Description
This PR adds autoscroll to the new moderation flow if a category is selected from a section and the section's view is hidden. It also fixes a spacing issue caught by @mplorentz, at the top and bottom of the scrollview when the view is scrolled.

## How to test
1. Build the app
2. Click the side menu
3. Click the Settings on the side menu
4. Scroll down to turn on the toggle for new moderation flow
5. Go back to the Feed screen by clicking the feed tab at the bottom of the screen
6. Find an account, Click on the 3 dots at the top right of the account.
7. Select `Flag this user` option. 
8. Select an option from the categories displayed.
9. Select an option on how to send the flag.
10. Select an option on whether to mute a user or not.
11. Click send.
12. Please confirm that:
    - The view scales to a new section if an option is picked from the previous section and the new section's view is not visible.

## Screenshots/Video

https://github.com/user-attachments/assets/c9092949-8c58-4a38-a289-9aa7001a618a

### CODE REVIEW

**Comment by joshuatbrown** [comment]
👀

**Review by joshuatbrown — COMMENTED** [review]
Great work so far. I love the animation.

File: Nos/Views/Moderation/ContentFlagView.swift [hunk 1]
```diff
@@ -15,6 +15,12 @@ struct ContentFlagView: View {
 
     @State private var flagCategories: [FlagOption] = []
 
+    /// Controls whether the "Send Section" should slide in from the left.
```
[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/ContentFlagView.swift:18]: Yes, that is what it means but it is better I make it clearer so it is easily understandable.

File: Nos/Views/Moderation/ContentFlagView.swift [hunk 2]
```diff
@@ -115,6 +130,25 @@ var flagSuccessView: some View {
...
+    proxy: ScrollViewProxy
+) {
+    withAnimation(.easeInOut(duration: 0.5)) {
+        shouldSlideIn.wrappedValue = true
+    }
+
+    withAnimation(.easeInOut(duration: 0.5)) {
+        proxy.scrollTo(targetSectionID, anchor: .bottom)
+    }
+}
```
[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/ContentFlagView.swift:150]: I have been thinking about this too. Thanks for your suggestion!!

File: Nos/Views/Moderation/UserFlagView.swift [hunk 3]
```diff
@@ -16,6 +16,18 @@ struct UserFlagView: View {
 
     @State private var flagCategories: [FlagOption] = []
 
+    /// Controls whether the "Send Section" should slide in from the left.
```
[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/UserFlagView.swift:19]: Maybe `animateSendSection`

File: Nos/Views/Moderation/UserFlagView.swift [hunk 4]
```diff
@@ -67,44 +78,63 @@ struct UserFlagView: View {
...
+                    if selectedSendOptionCategory != nil {
+                        FlagOptionPicker(
+                            previousSelection: .constant(nil),
+                            currentSelection: $selectedVisibilityOptionCategory,
+                            options: FlagOption.flagUserVisibilityOptions,
+                            title: String(localized: .localizable.flagUserMuteCategoryTitle),
+                            subtitle: nil
+                        )
+                        .id(visibilitySectionID)
+                        .offset(x: slideInVisibilitySection ? 0 : -UIScreen.main.bounds.width)
```
[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/UserFlagView.swift:118]: Nice pointing this out. I will check if there are other alternatives.

[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/UserFlagView.swift]: The geometry ready was a nice alternative.

**Comment by joshuatbrown** [comment]
👀

**Review by joshuatbrown — COMMENTED** [review]
One additional concern: I'm not seeing the view autoscroll when flagging a note (content). I do see it autoscroll when I flag a user. I did a little digging and it looks like we're not setting `.id(sendSectionID)` in the ContentFlagView like we do in the UserFlagView. Maybe that's it. This is looking great, though!

File: Nos/Views/Moderation/UserFlagView.swift [hunk 5]
```diff
@@ -70,44 +81,59 @@ struct UserFlagView: View {
...
+                        if selectedSendOption != nil {
+                            FlagOptionPicker(
+                                previousSelection: .constant(nil),
+                                currentSelection: $selectedVisibilityOption,
+                                options: FlagOption.flagUserVisibilityOptions,
+                                title: String(localized: .localizable.flagUserMuteCategoryTitle),
+                                subtitle: nil
+                            )
+                            .id(visibilitySectionID)
+                            .offset(x: animateVisibilitySection ? 0 : -UIScreen.main.bounds.width)
```
[the remark this answers is not in the document]
Comment (pelumy) [reply Nos/Views/Moderation/UserFlagView.swift:122]: 🤦‍♀️ Thanks for catching these!

**Comment by pelumy** [comment]
Thanks for catching all these @joshuatbrown. They have all been fixed.

**Comment by joshuatbrown** [comment]
👀

File: Nos/Views/Moderation/UserFlagView.swift [hunk 6]
```diff
@@ -70,44 +81,59 @@ struct UserFlagView: View {
...
+                                title: String(localized: .localizable.flagUserMuteCategoryTitle),
+                                subtitle: nil
+                            )
+                            .id(visibilitySectionID)
+                            .offset(x: animateVisibilitySection ? 0 : -geometry.size.width)
+                        }
+                    }
+                    .padding()
+                    .onChange(of: selectedFlagOption) {
+                        /// Animates the sliding effect and scrolls to the bottom of the specified section.
```
Comment (joshuatbrown) [inline Nos/Views/Moderation/UserFlagView.swift:127]: I think we could do without these comments (here and the one below, on line 131), but I'll leave it up to you. If you like them, they're fine with me.

**Review by joshuatbrown — APPROVED** [review]
Perfect! Great work on this!

File: Nos/Views/Moderation/UserFlagView.swift (hunk 6 above)
Comment (pelumy) [reply Nos/Views/Moderation/UserFlagView.swift:127]: Thanks @joshuatbrown , I just wanted to be sure anyone reading it understands the type of animation being added 😃

**Pull request merged**